src/parser/ComponentScopeResolver.ts

```diff
@@ -0,0 +1,78 @@
+import { ComponentDefinition } from "../componentprocessor";
+import * as Stmt from "./Statement";
+
+export class ComponentScopeResolver {
+    private readonly excludedNames: string[] = ["init"];
+
+    /**
+     * @param componentMap Component definition map to reference for function resolution.
+     * @param parserLexerFn Function used to parse statements out of given components
+     */
+    constructor(
+        readonly componentMap: Map<string, ComponentDefinition>,
+        readonly parserLexerFn: (filenames: string[]) => Promise<Stmt.Statement[]> // TODO: Remove and just build here?
+    ) {}
+
+    /**
+     * Resolves the component functions in scope based on the extends hierarchy.
+     * @param component Instance of the component to resolve function scope for.
+     * @returns All statements in scope for the resolved component
+     */
+    public async resolve(component: ComponentDefinition): Promise<Stmt.Statement[]> {
+        return Promise.all(this.getStatements(component)).then(this.flatten.bind(this));
+    }
+
+    /**
+     * Takes a sequence of statement collections and flattens them into a
+     * single statement collection. This function assumes that the components
+     * given in the statement map are in order of hierarchy with the furthest
+     * inheriting component first.
+     * @param statementMap Statement collections broken up by component.
+     * @returns A collection of statements that have been flattened based on hierarchy.
+     */
+    private flatten(statementMap: Stmt.Statement[][]): Stmt.Statement[] {
+        let statements = statementMap.shift() || [];
+        let statementMemo = new Set(
+            statements.filter((_): _ is Stmt.Function => true).map(statement => statement.name.text)
```

Comment:
So the `filter` here is for two purposes that aren't clear here (you might know a better way to do this). `statements` is already of type `Statement[]`, however I need it to be "cast" as a Function since that is what contains the name property. The type gate was kinda nice because it filters out types that wouldn't fit but also does the type casting for me once it gets into `map()`.

Reply:
Oh derp!  Right, that makes sense.  Ignore me — I think I misread a line or two

/me whistles and backs away nonchalantly